sql/core/src/test/scala/org/apache/spark/sql/MiniSparkSQLClusterSuite.scala

```diff
@@ -0,0 +1,168 @@
+          args(0).asInstanceOf[BlockId],
+          args(1).asInstanceOf[File],
+          args(2).asInstanceOf[SerializerInstance],
+          args(3).asInstanceOf[Int],
+          compressStream = identity,
+          syncWrites = false,
+          args(4).asInstanceOf[ShuffleWriteMetrics]
+        )
+      }
+    })
+    when(diskBlockManager.createTempShuffleBlock()).thenAnswer(
+      new Answer[(TempShuffleBlockId, File)] {
+        override def answer(invocation: InvocationOnMock): (TempShuffleBlockId, File) = {
+          val blockId = new TempShuffleBlockId(UUID.randomUUID)
+          val file = File.createTempFile(blockId.toString, null, tempDir)
+          blockIdToFileMap.put(blockId, file)
+          temporaryFilesCreated.append(file)
+          (blockId, file)
+        }
+      })
+    when(diskBlockManager.getFile(any[BlockId])).thenAnswer(
+      new Answer[File] {
+        override def answer(invocation: InvocationOnMock): File = {
+          blockIdToFileMap.get(invocation.getArguments.head.asInstanceOf[BlockId]).get
+        }
+      })
+  }
+
+  override def afterEach(): Unit = {
+    Utils.deleteRecursively(tempDir)
+    blockIdToFileMap.clear()
+    temporaryFilesCreated.clear()
+  }
+
+  private def toUnsafeRow(row: Row, schema: Array[DataType]): UnsafeRow = {
+    val internalRow = CatalystTypeConverters.convertToCatalyst(row).asInstanceOf[InternalRow]
+    val converter = UnsafeProjection.create(schema)
+    converter.apply(internalRow)
+  }
+
+  /**
+   * Create a spark context with specified configuration, and the calls `f`
+   */
+  protected def withSparkConf(sparkConfs: (String, String)*)(f: (SparkContext) => Unit): Unit = {
+    val (keys, values) = sparkConfs.unzip
+    val conf = new SparkConf(true)
+      .setMaster("local[1]")
+      .setAppName("testing")
+
+    var sc: SparkContext = null
+    val env = SparkEnv.get
+    Utils.tryWithSafeFinally {
+      (keys, values).zipped.foreach(conf.set)
+      sc = new SparkContext(conf)
+      f(sc)
+    } {
+      sc.stop()
+      SparkEnv.set(env)
+    }
+  }
+
+  test("SPARK-10466 external sorting for UnsafeRow with spilling") {
+    withSparkConf(
+      ("spark.shuffle.sort.bypassMergeThreshold", "0"),
+      ("spark.shuffle.spill.initialMemoryThreshold", "1024"), // 1k
+      ("spark.shuffle.memoryFraction", "0.0001")) { sc =>
+
+      // create unsafe rows
+      val count = 10000
+      val unsafeRowIt = (1 to count).iterator.map { i =>
+        (i, toUnsafeRow(Row(i, i, i), Array(IntegerType, IntegerType, IntegerType)))
+      }
+
+      val serializer = new UnsafeRowSerializer(3)
+      val sorter = new ExternalSorter[Int, UnsafeRow, UnsafeRow](
+        None, None, Some(implicitly[Ordering[Int]]), Some(serializer))
+      sorter.insertAll(unsafeRowIt)
+      // Make sure it spilled
+      assert(sc.env.blockManager.diskBlockManager.getAllFiles().length > 0)
+
+      assert(sorter.writePartitionedFile(shuffleBlockId, taskContext, outputFile).sum > 0)
```

Comment:
Exception will be thrown here if we didn't change the `UnsafeRowSerializer` as above.